### ISSUE TITLE
Sharded matmul is slower when its output's unit dim lands at position [-2]

### ISSUE THREAD

**Issue by elogir** [tenstorrent/tt-mlir#8620]
## Summary

Two logically-equivalent sharded matmuls — same operand shapes, same shardings, differing only in where the unit dim sits in the input shape — produce different TTNN IR and run at substantially different speeds. The slow form has the unit dim at position [-2] of the rank-4 shape that feeds `ttnn.reduce_scatter`; the fast form has the unit dim somewhere else.

I suspect the slowdown comes from the lowering's rank-4 left-pad combined with TILE layout's tile-padding of the last two dims: a unit dim that lands at position [-2] would get rounded up to a full tile (32), inflating the on-device collective allocation by ~32× along that axis.

https://github.com/tenstorrent/tt-mlir/blob/f400e65a698eb4a5385be94397cfa4ff58f7443d/lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/ReduceScatterOpRewritePattern.cpp#L30-L35
https://github.com/tenstorrent/tt-mlir/blob/f400e65a698eb4a5385be94397cfa4ff58f7443d/lib/Dialect/TTNN/Utils/Utils.cpp#L267-L274

## Example

At the TTIR level a row-parallel matmul produces a rank-3 partial-sum and JAX emits a `ttir.all_reduce` on it:

```mlir
%2 = "ttir.dot_general"(%0, %1) ...
    : (tensor<32x1x7168xbf16>, tensor<7168x4096xbf16>) -> tensor<32x1x4096xbf16>
%3 = "ttir.all_reduce"(%2) <{cluster_axis = 1, reduce_type = sum}>
    : (tensor<32x1x4096xbf16>) -> tensor<32x1x4096xbf16>
```

The TTIR→TTNN rewrite (`AllReduceReshapeOpRewritePattern.cpp` / `ReduceScatterOpRewritePattern.cpp`) lowers the all-reduce into `reduce_scatter` + `all_gather`. The `ttnn.matmul` itself is already collapsed to rank-2; the **unit dim only resurfaces in the reshape that precedes `ttnn.reduce_scatter`**, and its position there mirrors the user's original rank-3 input:

```mlir
// rank3_unit_S  — input [B=32, S=1, D] → reshape lands S=1 at [-2]:
%4 = "ttnn.matmul"(%3, %2) ... : (tensor<32x7168xbf16>, tensor<7168x4096xbf16>)
                              -> tensor<32x4096xbf16>
%5 = "ttnn.reshape"(%4) <{shape = [1, 32, 1, 4096]}>
                              : tensor<32x4096xbf16> -> tensor<1x32x1x4096xbf16>
%6 = "ttnn.reduce_scatter"(%5) ...
                              : tensor<1x32x1x4096xbf16> -> tensor<1x32x1x2048xbf16>

// rank3_S_leading — input [S=1, B=32, D] → reshape lands S=1 at [-3]:
%4 = "ttnn.matmul"(%3, %2) ... : (tensor<32x7168xbf16>, tensor<7168x4096xbf16>)
                              -> tensor<32x4096xbf16>
%5 = "ttnn.reshape"(%4) <{shape = [1, 1, 32, 4096]}>
                              : tensor<32x4096xbf16> -> tensor<1x1x32x4096xbf16>
%6 = "ttnn.reduce_scatter"(%5) ...
                              : tensor<1x1x32x4096xbf16> -> tensor<1x1x32x2048xbf16>
```

The only difference is the reshape's target shape: whether `S=1` lands at position [-2] or [-3]. The performance gap between the two forms is what triggered this report; the TILE-padding mechanism described in the Summary is my best guess at why.

## Reproducer

```python
import os, time
os.environ["CONVERT_SHLO_TO_SHARDY"] = "1"
import jax, jax.numpy as jnp, numpy as np
from jax.sharding import Mesh, NamedSharding, PartitionSpec

jax.config.update("jax_platforms", "tt")
B, S, D, DOUT = 32, 1, 14336, 4096
mesh = Mesh(np.array(jax.devices()), ("model",))

def matmul(x, w): return jnp.matmul(x, w)

def run(name, x_shape, x_spec):
    x = jax.device_put(jnp.ones(x_shape, jnp.bfloat16),
                       NamedSharding(mesh, x_spec))
    w = jax.device_put(jnp.ones((D, DOUT), jnp.bfloat16),
                       NamedSharding(mesh, PartitionSpec("model", None)))
    c = jax.jit(matmul, compiler_options={
        "export_path": "ir", "export_model_name": name}).lower(x, w).compile()
    for _ in range(5): jax.block_until_ready(c(x, w))
    t = time.perf_counter()
    for _ in range(100): out = c(x, w)
    jax.block_until_ready(out)
    print(f"{name}: {(time.perf_counter()-t)/100*1000:.3f} ms")

with mesh:
    run("rank3_unit_S",    (B, S, D), PartitionSpec(None, None, "model"))
    run("rank3_S_leading", (S, B, D), PartitionSpec(None, None, "model"))
```

Output:
```
rank3_unit_S: 2.078 ms
rank3_S_leading: 0.739 ms
```

Known workaround: transposing right before doing the matmul works and matches performance.
Note that I measured similar performance degradation on a full-sized LLM (Llama 3B/8B) with optimization level 1 or 2 and trace replay enabled.

## Possible resolution

A collective on a tensor with a unit (or otherwise small) dim should cost what the logical data movement implies. Frontends shouldn't have to transpose or collapse unit dims around every sharded matmul output just to dodge an internal padding decision.

Tested with `pjrt-plugin-tt==1.2.0.dev20260526002843`

**Comment by odjuricicTT** [comment]
@elogir This is expected behaviour for tenstorrent software and hardware. Reason for this is out tilized tensor layout https://docs.tenstorrent.com/tt-metal/latest/tt-metalium/tt_metal/advanced_topics/tiles.html

TLDR:
The last two dims always get padded to a multiple of 32. So you generally want to avoid having 1 in the last two dims.

In the above case the difference is in:
```
%5 = "ttnn.reshape"(%4) <{shape = [1, 32, 1, 4096]}>
                              : tensor<32x4096xbf16> -> tensor<1x32x1x4096xbf16>
```
vs
```
%5 = "ttnn.reshape"(%4) <{shape = [1, 1, 32, 4096]}>
                              : tensor<32x4096xbf16> -> tensor<1x1x32x4096xbf16>
```

Since the first reshape changes the tilized dims (last two), there is actually a lot of padding / unpadding that needs to happen under the hood which is. costly.
The second reshape is basically free, it's a no op.



If this answers you question, please close the issue.

**Comment by elogir** [comment]
The thing I was really after is whether this could be handled compiler-side rather than by the frontend. Since the two forms are logically equivalent and differ only in where the size-1 dim lands, would it be feasible for the lowering to detect that pattern and reorder so unit dims stay out of the last two (tiled) positions? That's effectively the transpose workaround, just done automatically.

If it's out of scope I'm happy to close, but it might be worth tracking as an enhancement.
